File: frontend/Customer_app/lib/features/restaurant/domain/entities/restaurant.dart
import 'package:flutter/material.dart';

import 'menu_category.dart';

/// Domain entity for a restaurant.
///
/// Temporal fields are parsed from the raw DTO strings into `TimeOfDay` and
/// `DateTime`; nullable defaults are applied by the mapper.
@immutable
class Restaurant {
  const Restaurant({
    required this.id,
    required this.name,
    this.description,
    this.address,
    this.city,
    this.isOpen,
    this.averageDeliveryTime,
    this.rating,
    this.imageUrl,
    this.logoUrl,
    this.coverImageUrl,
    this.cuisine,
    this.latitude,
    this.longitude,
    required this.active,
    this.openingTime,
    this.closingTime,
    this.createdAt,
    this.updatedAt,
    this.categories = const [],
  });

  final String id;
  final String name;
  final String? description;
  final String? address;
  final String? city;
  final bool? isOpen;
  final int? averageDeliveryTime;
  final double? rating;
  final String? imageUrl;
  final String? logoUrl;
  final String? coverImageUrl;
  final String? cuisine;
  final double? latitude;
  final double? longitude;
  final bool active;
  final TimeOfDay? openingTime;
  final TimeOfDay? closingTime;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final List<MenuCategory> categories;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Restaurant &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() =>
      'Restaurant(id: $id, name: $name, rating: $rating, open: $isOpen)';
}

File: backend/restaurant/src/main/java/com/service/restaurant/entity/Restaurant.java
package com.service.restaurant.entity;

import jakarta.persistence.*;
import lombok.*;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.time.LocalDateTime;
import java.time.LocalTime;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Entity
@Table(
        name = "restaurants",
        indexes = {
                @Index(name = "idx_restaurant_city", columnList = "city"),
                @Index(name = "idx_restaurant_owner", columnList = "owner_id"),
                @Index(name = "idx_restaurant_active", columnList = "active"),
                @Index(name = "idx_restaurant_deleted", columnList = "deleted"),
                @Index(name = "idx_restaurant_cuisine", columnList = "cuisine"),
                @Index(name = "idx_restaurant_name", columnList = "name"),
                @Index(name = "idx_restaurant_customer_city", columnList = "active, deleted, city")
        }
)
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Restaurant {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false)
    private String name;

    private String description;

    @Column(nullable = false)
    private String phone;

    @Column(nullable = false)
    private String address;

    private String city;

    private Double latitude;

    private Double longitude;

    @Column(nullable = false)
    private Boolean open;

    private Integer averageDeliveryTime;

    private Double rating;

    private String imageUrl;

    private String logoUrl;

    private String coverImageUrl;

    private String cuisine;

    private LocalTime openingTime;

    private LocalTime closingTime;

    @OneToMany(
            mappedBy = "restaurant",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    @Builder.Default
    private List<MenuCategory> categories = new ArrayList<>();

    @Builder.Default
    @Column(nullable = false, columnDefinition = "boolean default true")
    private boolean active = true;

    @Builder.Default
    @Column(nullable = false, columnDefinition = "boolean default false")
    private boolean deleted = false;

    @Column(name = "owner_id", nullable = false)
    private UUID ownerId;

    @CreatedDate
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;
}
